prov/efa/src/rxr/rxr_ep.c

```diff
@@ -1429,6 +1439,21 @@ void rxr_ep_progress_internal(struct rxr_ep *ep)
 		}
 	}
 
+	/*
+	 * Send read requests until finish or error encoutered
```

Comment:
Before this patch, we tracked the number of outstanding requests with `ep->tx_pending` and never tried to post a request when the queue is full. However, it looks like the read requests count against that limit as well. What could happen here is the work queue is almost filled while sending data segments and then we try to issue a read in this loop. The EFA libibverbs provider will return ENOMEM since the queue is full and we'll write an error. Instead, we can break out of the progress thread when the work queue is full.

Reply:
Thank you! In new revision, I added call to rxr_ep_inc_tx_pending() after fi_read() and call to rxr_ep_dec_tx_pending() after read completion. Meanwhile, in rxr_ep_progress_internal(), I added the check of ep->tx_pending and ep->max_outstanding_tx. These changes should be enough to address this issue?

Reply:
Yes, that will work.